### DISCUSSION TITLE
Add message redelivery counter

### DISCUSSION THREAD

**Discussion by mikle-a** [rabbitmq/rabbitmq-server#9079] [Other]
### Is your feature request related to a problem? Please describe.

Currently I am using simple queues to dispatch heavy jobs between worker instances of my java app. Sometimes, jobs processing results in unexpected errors causing instance to crash without any capability to reject source message before dying. In this case 
message is simply redelivered to another instance which most probably will crash as well. As far as I see, `Envelop` already provides convenient method `isRedelivered()` to resolve whether it is not the first attempt already, but it would be great to get particular attempt counter as well. It will help an application to check the counter before processing and reject it to DLQ if it exceeds some kind of a threshold.

### Describe the solution you'd like

Just some additional header like `x-retry-count` passed from message broker to a consumer will be quite enough.

### Describe alternatives you've considered

I know that similar header is already added to a message when it is returned from DLQ back to processing, but as I said in the beginning, sometimes it is not possible to reject inbound message

**Comment by michaelklishin** [comment]
Usually such counters are used to
cap the number of retries. [Quorum queues already allow you to configure a delivery limit](https://rabbitmq.com/quorum-queues.html#poison-message-handling).

**Reply by mikle-a** [reply]
Yes, from the functional side it is exactly what is needed. However, as I understand, quorum queues are highly focused on the reliability sacrificing the performance, which is not applicable in all cases. So, it would be great to have the same functionality, but for the classic queues as well :)

**Comment by ikavgo** [comment]
Similar to this: https://github.com/rabbitmq/rabbitmq-server/issues/2569

**Comment by mikle-a** [comment]
I'm sure I'm not alone, please do not hesitate to press 👍 if you need this functionality as well

**Reply by sujalmandal** [reply]
Having this information in the Envelope can be great to drive some logic other than just publishing to DLQ. Would like to see this one.

**Comment by lhoguin** [comment]
I plan to add this for the version after 3.13 (which should be 4.0). It will only work for CQv2.
